test/functional/change_stream.test.js

```diff
@@ -2597,233 +2599,236 @@ describe('Change Streams', function() {
-          getMore: (function*() {
-            yield { numDocuments: 1, postBatchResumeToken: false };
-          })()
+    });
+
+    afterEach(function(done) {
+      client.close(done);
+    });
+
+    it('should work with events', {
+      metadata: { requires: { topology: 'replicaset', mongodb: '>=4.1.1' } },
+      test: function(done) {
+        const changeStream = coll.watch([], { startAfter });
+        changeStream.once('change', change => {
+          expect(change).to.containSubset({
+            operationType: 'insert',
+            fullDocument: { x: 2 }
+          });
+          changeStream.close();
         });
-        let token;
-        const resumeAfter = manager.resumeToken();
+        changeStream.once('close', () => setTimeout(done));
+      }
+    });
 
-        return manager
-          .ready()
-          .then(() => {
-            const changeStream = manager.makeChangeStream({ resumeAfter });
-            let counter = 0;
-            changeStream.cursor.on('response', () => {
-              if (counter === 1) {
-                token = changeStream.resumeToken;
-              }
-              counter += 1;
-            });
+    it('should work with callbacks', {
+      metadata: { requires: { topology: 'replicaset', mongodb: '>=4.1.1' } },
+      test: function(done) {
+        const changeStream = coll.watch([], { startAfter });
+        exhaust(changeStream, (err, bag) => {
+          expect(err).to.not.exist;
+          const finalOperation = bag.pop();
+          expect(finalOperation).to.containSubset({
+            operationType: 'insert',
+            fullDocument: { x: 2 }
+          });
+          changeStream.close(done);
+        });
+      }
+    });
 
-            // Note: this is expected to fail
-            return changeStream.next();
-          })
-          .then(
-            () => manager.teardown(),
-            err => manager.teardown(err)
-          )
-          .then(() => {
-            expect(token).to.deep.equal(resumeAfter);
+    // 17. $changeStream stage for ChangeStream started with startAfter against a server >=4.1.1
+    // that has not received any results yet
+    // - MUST include a startAfter option
+    // - MUST NOT include a resumeAfter option when resuming a change stream.
+    it('$changeStream that has not received results must include startAfter and not resumeAfter', {
+      metadata: { requires: { topology: 'replicaset', mongodb: '>=4.1.1' } },
+      test: function(done) {
+        const changeStream = coll.watch([], { startAfter });
+        tryNext(changeStream, (err, change) => {
+          expect(err).to.not.exist;
+          expect(change).to.containSubset({
+            operationType: 'insert',
+            fullDocument: { x: 2 }
+          });
+          triggerResumableError(changeStream);
+          tryNext(changeStream, (err, change) => {
+            expect(err).to.exist;
+            expect(change).to.not.exist;
+            const retry = setInterval(
```

Comment:
can you explain why we need a retry loop here? I'm trying to determine if this is the best way to test this. The prose seems to require that we verify that the outgoing `aggregate` command has a set of expected values (or isn't sending some value), so it seems like we should be using command monitoring here, rather than retrying an indefinite amount of times, and then directly inspecting the cursor's operation.

Reply:
The retry is just there because `tryNext` errors until the topology has reconnected after a resumable error is encountered. On reflection, there are definitely more elegant ways to do that, I'm open to suggestions! :)

Reply:
In this case I think you don't want `tryNext`, but instead to call `cursor.next(..)`. The idea is that the stream is not broken, but you are using the stepwise `tryNext`.

Regarding the command assertions, let's use command monitoring rather than depending on "private implementation". You'll want to attach a listener to the client's `commandStarted` event, and also pass `monitorCommands: true` to the client's constructor in `newClient`. `changeStream.cursor.operation.pipeline` might change at some point, so it's brittle to depend on that being there.

Reply:
I refactored prose tests 17 and 18 to use command monitoring, and events rather than iteration. That unblocks this with respect to [NODE-2548](https://jira.mongodb.org/browse/NODE-2548); I'll go back and implement these tests with promises/callbacks as part of that work.

I also fixed a bug causing prose test 17 to fail in a separate commit.